beard/clustering/blocking.py

```diff
@@ -97,54 +144,53 @@ def _validate(self, X, blocks):
 
-    def _fit(self, X, y, blocks):
-        """Fit base clustering estimators on X."""
-        self.labels_ = -np.ones(len(X), dtype=np.int)
-        offset = 0
+    def _block(self, X, y, blocks, affinity):
+        """Chop the training data into smaller chunks.
 
-        for b in np.unique(blocks):
-            # Select data from block
-            mask = (blocks == b)
-            X_mask = X[mask, :]
+        A chunk is demarcated by the corresponding block. Each chunk contains
+        only the training examples relevant to given block.
+
+        Returns
+        -------
+        :returns: array
+            Triples in form of (block, X, y) where X and y are the training
+            examples for given block.
+        """
+        unique_blocks = np.unique(blocks)
 
+        chunk = []
+        for single_block in unique_blocks:
+            mask = (blocks == single_block)
+            X_mask = X[mask, :]
             if y is not None:
                 y_mask = y[mask]
             else:
                 y_mask = None
-
-            if self.affinity == "precomputed":
+            if affinity == "precomputed":
                 X_mask = X_mask[:, mask]
+            chunk.append((single_block, X_mask, y_mask))
+        return chunk
+
+    def _fit(self, X, y, blocks):
+        """Fit base clustering estimators on X."""
+        self.labels_ = -np.ones(len(X), dtype=np.int)
+        offset = 0
 
-            # Fit a clusterer on the selected data
-            if len(X_mask) == 1:
-                clusterer = _SingleClustering()
-            elif self.fit_:
-                clusterer = clone(self.base_estimator)
-            elif self.partial_fit_:
-                if b in self.clusterers_:
-                    clusterer = self.clusterers_[b]
-                else:
-                    clusterer = clone(self.base_estimator)
-
-            if self.fit_ or not hasattr(clusterer, "partial_fit"):
-                try:
-                    clusterer.fit(X_mask, y=y_mask)
-                except TypeError:
-                    clusterer.fit(X_mask)
-            elif self.partial_fit_:
-                try:
-                    clusterer.partial_fit(X_mask, y=y_mask)
-                except TypeError:
-                    clusterer.partial_fit(X_mask)
-
-            self.clusterers_[b] = clusterer
-
-            # Save predictions
-            pred = np.array(clusterer.labels_)
-            mask_unknown = (pred == -1)
-            pred[~mask_unknown] += offset
-            self.labels_[mask] = pred
-            offset += np.max(clusterer.labels_) + 1
+        result = (Parallel(n_jobs=self.n_jobs)
+                  (delayed(_parallel_fit)(self.fit_, self.partial_fit_,
+                                          self.base_estimator,
+                                          self.clusterers_, *b) for b in
+                   self._block(X, y, blocks, self.affinity)))
+
+        for bs in result:
+            for b, clusterer in bs.items():
```

Comment:
provided `_parallel_fit` returns a tuple, we can now remove this double loop and do `for b, clusterer in results:`